### ISSUE TITLE
provide a `Map<N, Point> getLocations()` method on `LayoutModel`

### ISSUE THREAD

**Issue by jrtom** [jrtom/jung#184]
We ran into some problems with this described in the comments to PRs #180 and #182.  

I have a couple of proposed solutions.

Fundamentally, we have a `Set` of nodes (from the graph) and you have a `Function` that you can apply to an element to get its value.  You just need a way to combine them, using a live read of the graph each time.

Option (1): Guava's [`Maps.asMap(Set, Function)`](http://google.github.io/guava/releases/21.0/api/docs/com/google/common/collect/Maps.html#asMap-java.util.Set-com.google.common.base.Function-)

This gives you a map that's based on a live view of the graph nodes.  Code would look something like:

```
default public  Map<N, Point> getLocations() {
  return Maps.asMap(getGraph().nodes(), LayoutModel::get); 
}
```

and that should do what you want, i.e., rely on the `Graph` as the source of truth.

`Maps.asMap` operates on a Guava `Function`, rather than a Java 8 `Function`, which might present a minor problem, but I think we could work it out.

Option (2): Stream

```
default public  Map<N, Point> getLocations() {
  return getGraph().nodes().stream().collect(Function.identity(), LayoutModel::get); 
}
```

I _think_ that this also treats the nodes of the graph as a live view, but I'm less familiar with the `Stream` idiom.  @jbduncan could probably comment on whether this would work as hoped.  :)

@tomnelson any thoughts?

**Comment by jbduncan** [comment]
Hi @jrtom. I believe that both of the options you describe would indeed treat the underlying graph's nodes as a live view, but whereas option (1) would return a live map, option (2) would instead return an eagerly-evaluated map - a "snapshot", if you will. So it depends on whether you and @tomnelson expect the map returned from `getLocations()` to change on-the-fly when the underlying graph's nodes also changes, or if you expect a static, never-changing map to be returned instead (in which case it may make more sense to use Guava's `ImmutableMap.toImmutableMap` collector over the built-in `Collectors.toMap`).

I hope this helps. :)

**Comment by tomnelson** [comment]
I really like these, but I appear to like them a lot more than the compiler does! 
The compiler was not happy with the method reference as a Function.

I fear that the second approach collects the mappings into a newly created Map, doing extra work that might be a problem with large graphs. 

I find that I can do this, and make the compiler happy:

`public interface LayoutModel<N> extends Function<N, Point>, com.google.common.base.Function<N, Point> {
   Map<N, Point> getLocations();
`

(adding a guava Function interface to the LayoutModel, just for this method).
Then, in AbstractLayoutModel, implement it like this:

`  public Map<N, Point> getLocations() {
    return Maps.asMap(getGraph().nodes(), this);
  }
`

I think that will make the earlier PR branch work as desired, preserve the use of getLocations() method for iteration over the node/point mappings, and leave the final implementation of how to best make the mapping as follow on work (or at least give us all a place to try things).

**Comment by jrtom** [comment]
Implemented in #185.
